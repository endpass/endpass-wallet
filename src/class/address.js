import web3 from 'web3';
import { WALLET_TYPE, HARDWARE_WALLET_TYPE } from '@/constants';
import { TrezorWallet } from '@/class';

const { isAddress, hexToBytes, bytesToHex, toChecksumAddress } = web3.utils;

// A wallet based on an Ethereum address string, to be used in place of
// ethereumjs-wallet when only the address is available
export default class Address {
  constructor({ address, info }) {
    const addressString = `0x${address.replace(/^0x/, '')}`;

    if (!isAddress(addressString)) {
      throw new Error('Not a valid Ethereum address');
    }

    this.address = hexToBytes(addressString);
    this.isPublic = true;
    this._privKey = null;
    this._pubKey = null;
    this.privKey = null;
    this.pubKey = null;

    if (info) {
      this.info = info;

      const isHardware = Object.values(HARDWARE_WALLET_TYPE).includes(
        info.type,
      );

      if (isHardware) {
        this.isPublic = false;
      }

      switch (info.type) {
        case WALLET_TYPE.TREZOR:
          this.signStrategy = TrezorWallet;
          break;

        default:
          this.signStrategy = TrezorWallet;
          break;
      }
    }
  }

  getAddress() {
    return this.address;
  }

  getAddressString() {
    return bytesToHex(this.address);
  }

  getChecksumAddressString() {
    return toChecksumAddress(this.getAddressString());
  }

  // (Message) => Promise<SignedMessage>
  sign(message) {
    return this.signStrategy.sign(message, this.info.index);
  }

  // (Transaction) => Promise<SignedTrx>
  signTransaction(transaction) {
    return this.signStrategy.signTransaction(transaction, this.info.index);
  }
}
