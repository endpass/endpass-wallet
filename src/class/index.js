export { providerFactory } from './provider';
export { default as Transaction } from './Transaction';
export {
  default as TransactionFactory,
} from './transaction/TransactionFactory';
export { default as EventEmitter } from './EventEmitter';
export { NotificationError } from './error';
export { default as Address } from './address';
export { default as Wallet } from './wallet';
export { ENSResolver } from './ens';
export { Token } from './Token';
export { default as ERC20Token } from './erc20';
export { default as TrezorWallet } from './hardwareWallet/TrezorWallet';
export { default as LedgerWallet } from './hardwareWallet/LedgerWallet';
