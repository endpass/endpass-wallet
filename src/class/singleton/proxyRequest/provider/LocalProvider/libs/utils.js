import regPath from 'path-to-regexp';

const utils = {
  getRouteKey(routesRegexp, routePath) {
    return Object.keys(routesRegexp).find(key => {
      const re = routesRegexp[key];
      return re.test(routePath);
    });
  },
  getPath(serverUrl, url) {
    const levels = url.split(serverUrl);
    const keysPath = levels.pop() || '';
    const ret = keysPath[0] === '/' ? keysPath.substring(1) : keysPath;

    return ret;
  },
  getPathArgs(routesRegexp, key, routePath) {
    const keys = regPath.parse(key);
    const re = routesRegexp[key];
    re.lastIndex = 0; // drop exec position

    const values = re.exec(routePath);
    return keys.reduce((map, item, index) => {
      if (item && item.name) {
        map[item.name] = values[index];
      }
      return map;
    }, {});
  },
};

export default utils;
