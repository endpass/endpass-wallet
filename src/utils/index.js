import store from '../store';

export { default as http } from './http';

export const kebabToCamel = input =>
  input.replace(/-([a-z])/g, g => g[1].toUpperCase());

export const getInitializedValueFromStore = (module, field) => {
  return new Promise(resolve => {
    const value = module[field];

    if (value === null) {
      const unwatch = store.watch(
        () => module[field],
        newValue => {
          unwatch();
          resolve(newValue);
        },
      );
    } else {
      resolve(value);
    }
  });
};
