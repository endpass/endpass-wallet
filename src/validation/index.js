import Vue from 'vue';
import { default as VeeValidate, Validator } from 'vee-validate';
import address from './address';
import privateKey from './privateKey';
import publicKey from './publicKey';
import seedPhrase from './seedPhrase';
import ensOrAddress from './ensOrAddress';
import hex from './hex';

Validator.extend('address', address);
Validator.extend('ensOrAddress', ensOrAddress, {
  immediate: false,
});
Validator.extend('private_key', privateKey);
Validator.extend('public_key', publicKey);
Validator.extend('seed_phrase', seedPhrase);
Validator.extend('hex', hex);

Vue.use(VeeValidate, { inject: false });
