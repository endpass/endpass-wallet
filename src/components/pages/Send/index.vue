<template>
  <div class="app-page send-page">
    <div class="section is-narrow">
      <div class="container is-narrow">
        <div class="card app-card main-app-card">
          <div class="card-header">
            <h1 class="card-header-title">Send ETH</h1>
          </div>
          <div class="card-content">
            <transaction-form
              :transaction="transaction"
              @submit="handleTransactionSend"
            />

            <div
              v-if="transactionHash"
              class="transaction-status message is-success"
              data-test="transaction-status"
            >
              <div class="message-header">
                <p>Transaction Sent!</p>
              </div>
              <div class="message-body">
                <p>Your transaction has been broadcast to the network. It may take a few minutes before the transaction is confirmed.</p>
                <p class="label">Transaction Id</p>
                <p class="code">{{ transactionHash }}</p>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
    <transaction-modal
      v-if="isWaitingConfirm"
      :transaction="transaction"
      @confirm="confirmTransaction"
      @close="cancelTransaction"
    />
    <password-modal
      v-if="isTransactionConfirmed"
      @confirm="sendConfirmedTransaction"
      @close="cancelTransaction"
    />
  </div>
</template>

<script>
import { mapState, mapActions } from 'vuex';
import { TransactionFactory } from '@/class';
import VForm from '@/components/ui/form/VForm.vue';
import VRadio from '@/components/ui/form/VRadio.vue';
import VSelect from '@/components/ui/form/VSelect';
import VInput from '@/components/ui/form/VInput.vue';
import VSpinner from '@/components/ui/VSpinner';
import VInputAddress from '@/components/ui/form/VInputAddress.vue';
import VButton from '@/components/ui/form/VButton.vue';
import AccountChooser from '@/components/AccountChooser';
import TransactionModal from '@/components/modal/TransactionModal';
import PasswordModal from '@/components/modal/PasswordModal';
import privatePage from '@/mixins/privatePage';
import { getShortStringWithEllipsis } from '@/utils/strings';
import TransactionForm from './TransactionForm';

const defaultTx = {
  tokenInfo: null,
  gasPrice: '40',
  gasLimit: '22000',
  value: '0',
  nonce: 0,
  to: '',
  data: '0x',
};

export default {
  data: () => ({
    transaction: { ...defaultTx },
    transactionHash: null,
    isSending: false,
    isWaitingConfirm: false,
    isTransactionConfirmed: false,
  }),

  computed: {
    ...mapState({
      activeAddress: state => state.accounts.address.getChecksumAddressString(),
      activeNet: state => state.web3.activeNet,
    }),
  },

  watch: {
    async activeAddress() {
      this.transactionHash = null;
      this.transaction.nonce = await this.getNextNonce();
    },

    async activeNet() {
      this.transactionHash = null;
      this.transaction.nonce = await this.getNextNonce();
    },
  },

  methods: {
    ...mapActions('transactions', ['sendTransaction', 'getNextNonce']),
    ...mapActions('gasPrice', ['getGasPrice']),

    handleTransactionSend() {
      this.isWaitingConfirm = true;
    },

    cancelTransaction() {
      this.isWaitingConfirm = false;
      this.isTransactionConfirmed = false;
    },

    async confirmTransaction() {
      this.isWaitingConfirm = false;
      this.isTransactionConfirmed = true;
    },

    async sendConfirmedTransaction(password) {
      this.isSending = true;

      Object.assign(this.transaction, {
        from: this.activeAddress,
        networkId: this.activeNet.id,
      });

      this.resetForm();
      this.isTransactionConfirmed = false;
      this.isWaitingConfirm = false;

      try {
        const hash = await this.sendTransaction({
          transaction: TransactionFactory.fromSendForm(this.transaction),
          password,
        });
        const shortHash = getShortStringWithEllipsis(hash);

        this.transactionHash = hash;

        this.$notify({
          title: 'Transaction Sent',
          text: `Transaction ${shortHash} sent`,
          type: 'is-info',
        });
      } catch (err) {
        this.transactionHash = null;
      } finally {
        this.isSending = false;
      }
    },

    async handleTransactionFormSubmit() {
      this.toggleTransactionModal();
    },

    async resetForm() {
      const nonce = await this.getNextNonce();

      this.transaction = {
        ...defaultTx,
        nonce,
      };
    },
  },

  async created() {
    this.transaction.nonce = await this.getNextNonce();
  },

  mixins: [privatePage],

  components: {
    VForm,
    VButton,
    VRadio,
    VSpinner,
    VInput,
    VInputAddress,
    VSelect,
    AccountChooser,
    TransactionModal,
    PasswordModal,
    TransactionForm,
  },
};
</script>

<style lang="scss">
.send-page {
  .send-amount {
    margin-top: 2em;
    margin-bottom: 2em;
  }

  .field-label {
    margin-bottom: 0;
  }
}
</style>
