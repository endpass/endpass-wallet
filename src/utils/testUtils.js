import { kebabCase } from 'lodash';
import { getFrom } from '@/utils/objects';

function convertDirectivesToAttrs(directives) {
  const attrs = {};

  if (directives) {
    directives.forEach(({ rawName, expression }) => {
      attrs[rawName] = expression;
    });
  }

  return attrs;
}

function convertListenersToAttrs(listeners) {
  const attrs = {};

  if (listeners) {
    Object.keys(listeners).forEach(key => {
      const handlers = [].concat(listeners[key]);
      const handler = handlers.find(h => {
        const name = getFrom(h, 'fns.name', 'name') || '';

        return name.includes('bound ');
      });

      if (handler) {
        attrs[`v-on:${key}`] = (handler.fns || handler).name;
      }
    });
  }

  return attrs;
}

function convertModelToAttrs(model) {
  return model ? { 'v-model': model.expression } : {};
}

function convertSlotToAttrs(slot) {
  return slot ? { slot } : {};
}

function generateElement(createElement, vnode) {
  return vnode.componentOptions
    ? generateElementFromComponent(createElement, vnode)
    : generateElementFromHTML(createElement, vnode);
}

function getChildrenElements(createElement, children) {
  return children
    ? children.map(child => generateElement(createElement, child))
    : undefined;
}

function generateElementFromComponent(
  createElement,
  { data, componentOptions },
) {
  const { attrs, directives, model } = data;
  const { tag, listeners, children } = componentOptions;
  const componentData = {
    ...data,
    attrs: {
      ...attrs,
      ...convertDirectivesToAttrs(directives),
      ...convertListenersToAttrs(listeners),
      ...convertModelToAttrs(model),
      class: data.staticClass,
    },
    hook: undefined,
    staticClass: undefined,
  };

  if (listeners) {
    componentData.on = Object.assign({}, listeners);
  }

  return createElement(
    tag,
    componentData,
    getChildrenElements(createElement, children),
  );
}

function generateElementFromHTML(createElement, vnode) {
  if (!vnode.data) {
    return vnode;
  }

  const { tag, data, children } = vnode;
  const { directives, on, model, attrs = {}, slot } = data;

  return createElement(
    tag,
    {
      ...data,
      attrs: {
        ...attrs,
        ...convertDirectivesToAttrs(directives),
        ...convertListenersToAttrs(on),
        ...convertModelToAttrs(model),
        ...convertSlotToAttrs(slot),
      },
    },
    getChildrenElements(createElement, children),
  );
}

export function generateStubs(Component) {
  return Object.values(Component.components).reduce((stubs, stubComponent) => {
    let elementName;

    if (stubComponent.name) {
      elementName = kebabCase(stubComponent.name);
    }

    return Object.assign(stubs, {
      [elementName]: {
        render(createElement) {
          return generateElementFromComponent(createElement, this.$vnode);
        },
      },
    });
  }, {});
}

export default {
  generateStubs,
};
