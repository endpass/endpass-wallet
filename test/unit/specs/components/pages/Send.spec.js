import Vue from 'vue';
import validation from '@/validation';
import Vuex from 'vuex';
import Web3 from 'web3';
import { shallow, createLocalVue, mount } from '@vue/test-utils';
import Notifications from 'vue-notification';
import { Transaction } from '@/class';
import { generateStubs } from '@/utils/testUtils';

import Send from '@/components/pages/Send.vue';

Transaction.prototype.getFullPrice = jest.fn().mockResolvedValue();

const localVue = createLocalVue();

localVue.use(Vuex);
localVue.use(Notifications);

describe('Send', () => {
  let actions;
  let store;
  let wrapper;
  const web3Instance = new Web3('https://mainnet.infura.io/');

  beforeEach(() => {
    store = new Vuex.Store({
      state: {
        price: {
          price: 400,
        },
        user: {
          settings: {
            fiatCurrency: 'USD',
          },
        },
        web3: {
          web3: web3Instance,
          activeCurrency: {
            name: 'ETH',
          },
          activeNet: {
            id: 1,
          },
        },
        connectionStatus: {
          isSyncing: false,
        },
      },
      actions,
      modules: {
        accounts: {
          namespaced: true,
          state: {
            address: {
              getChecksumAddressString() {
                return '0x9eceefdf3554e178a6549006f2c02163e63c9fd8';
              },
            },
            balance: '1000000000000000000',
            pendingTransactions: [
              {
                timestamp: 1524505925,
                from: '0x4bd5c3e7e4d6b3df23e9da5b42e5e4daa3d2579b',
                to: '0x7c59542b20002ed255598172cab48b86d865dfbb',
                hash:
                  '0x7fcb1e71def6d0d353251831f46d60401e6321b5e0b0b135085be4688ca2a9b1',
                value: 0.009979,
                input: '0x',
                success: true,
              },
            ],
            wallets: () => [],
            wallet: {
              isPublic: false,
            },
          },
          getters: {
            isPublicAccount: () => true,
          },
        },
        gasPrice: {
          namespaced: true,
          actions: {
            getGasPrice() {
              return new Promise(res =>
                res({
                  low: '1',
                  medium: '2',
                  high: '3',
                }),
              );
            },
          },
        },
        tokens: {
          namespaced: true,
          getters: {
            allCurrentAccountTokensWithNonZeroBalance: () => [
              {
                symbol: 'AAA',
                address: '0xB6eD7644C69416d67B522e20bC294A9a9B405B31',
                decimals: 8,
              },
            ],
          },
        },
        transactions: {
          namespaced: true,
          actions: {
            getNextNonce: jest.fn(),
            getNonceInBlock: jest.fn(),
          },
          getters: {
            addressesFromTransactions: () => [],
          },
        },
      },
    });
    wrapper = shallow(Send, { store, localVue });
  });

  describe('render', () => {
    beforeEach(() => {
      wrapper = shallow(Send, {
        store,
        localVue,
        stubs: generateStubs(Send),
      });
    });

    it('should render ens error', async () => {
      expect.assertions(1);

      wrapper.setData({
        ensError: 'foo',
      });

      await wrapper.vm.$nextTick();

      expect(wrapper.html()).toMatchSnapshot();
    });

    it('should render resolved ens address', async () => {
      expect.assertions(1);

      wrapper.setData({
        transaction: {
          to: '0x0',
        },
      });
      wrapper.setComputed({
        isEnsTransaction: true,
      });

      await wrapper.vm.$nextTick();

      expect(wrapper.html()).toMatchSnapshot();
    });
  });

  describe('behavior', () => {
    it('should validate data', async () => {
      const getFullPrice = jest.fn().mockResolvedValue('400');

      wrapper = mount(Send, {
        store,
        localVue,
        computed: {
          maxAmount: () => 2,
        },
      });

      const { errors } = wrapper.vm;

      wrapper.setData({
        transaction: {
          to: '',
          gasPrice: '900',
          gasLimit: '2200000000',
          data: 'asdfas',
          getFullPrice,
        },
        value: '2.222222222222222222222222',
      });

      await wrapper.vm.$validator.validateAll();

      expect(errors.first('gasPrice').includes('between')).toBeTruthy();
      expect(errors.first('gasLimit').includes('between')).toBeTruthy();
      expect(errors.first('price').includes('between')).toBeTruthy();
      expect(errors.first('value').includes('decimal')).toBeTruthy();
      expect(errors.first('data').includes('hex')).toBeTruthy();

      wrapper.setData({
        transaction: {
          ...wrapper.vm.transaction,
          to: '',
          gasPrice: '91',
          gasLimit: '22000',
          // prettier-ignore
          data: '0xa9059cbb000000000000000000000000b6ed7644c69416d67b522e20bc294a9a9b405b31000000000000000000000000000000000000000000000000000000003b9aca00'
        },
        value: '1.5',
      });

      await wrapper.vm.$validator.validateAll();

      expect(errors.has('gasPrice')).toBeFalsy();
      expect(errors.has('gasLimit')).toBeFalsy();
      expect(errors.has('price')).toBeFalsy();
      expect(errors.has('value')).toBeFalsy();
      expect(errors.has('data')).toBeFalsy();

      wrapper.setData({
        transaction: {
          ...wrapper.vm.transaction,
          to: '123',
          gasPrice: '-90',
          gasLimit: '-22000',
          data: '0x',
        },
        value: '-2.22222',
      });

      await wrapper.vm.$validator.validateAll();

      expect(errors.first('gasPrice').includes('numeric')).toBeTruthy();
      expect(errors.first('gasLimit').includes('numeric')).toBeTruthy();
      expect(errors.first('price').includes('between')).toBeTruthy();
      expect(errors.first('value').includes('between')).toBeTruthy();

      wrapper.setComputed({
        balance: '0',
      });

      wrapper.setData({
        estimateGasCost: '1000000000000000',
      });

      expect(errors.firstById('insufficientBalance')).toBeTruthy();

      wrapper.setComputed({
        balance: '1000000000000000000',
      });

      wrapper.setData({
        estimateGasCost: '1000000000000000',
      });

      expect(errors.firstById('insufficientBalance')).toBeFalsy();
    });

    it('should allow empty string in input', () => {
      wrapper = mount(Send, {
        store,
        localVue,
      });

      const inputIdArr = ['gasPrice', 'gasLimit', 'value'];

      inputIdArr
        .map(inputId => wrapper.find(`#${inputId}`))
        .forEach(inputElem => {
          inputElem.element.value = '';
          inputElem.trigger('input');
          expect(inputElem.element.value).toBe('');
        });
    });

    it('should sync amount and price input', () => {
      wrapper = mount(Send, {
        store,
        localVue,
      });

      const valueInput = wrapper.find('input#value');
      const priceInput = wrapper.find('input#price');

      valueInput.element.value = 2;
      valueInput.trigger('input');

      expect(wrapper.vm.price).toBe('800.00');

      priceInput.element.value = 1200;
      priceInput.trigger('input');

      expect(wrapper.vm.value).toBe('3.000000000000000000');
    });

    it('should set transaction value to the maximum amount', () => {
      wrapper.vm.$validator.validate = jest.fn();

      wrapper.vm.setMaxAmount();

      const maxAmount = wrapper.vm.maxAmount.toString();

      expect(wrapper.vm.transaction.value).toBe(maxAmount);
      expect(wrapper.vm.value).toBe(maxAmount);
    });

    it('should disable price input when zero eth price', () => {
      wrapper = shallow(Send, {
        store,
        localVue,
        stubs: generateStubs(Send),
      });

      const inputElem = wrapper.find('v-input#price');

      wrapper.setComputed({
        ethPrice: 0,
      });

      expect(inputElem.attributes().disabled).toBeTruthy();

      wrapper.setComputed({
        ethPrice: 1,
      });

      expect(inputElem.attributes().disabled).toBeFalsy();
    });

    it('should update user nonce when changing the network', async () => {
      wrapper.vm.getNextNonce = jest.fn().mockResolvedValueOnce('2');
      wrapper.setComputed({
        activeNet: { id: 2 },
      });
      await wrapper.vm.$nextTick();

      expect(wrapper.vm.$data.userNonce).toBe('2');

      wrapper.vm.getNextNonce.mockResolvedValueOnce('5');
      wrapper.setComputed({
        activeNet: { id: 3 },
      });
      await wrapper.vm.$nextTick();

      expect(wrapper.vm.$data.userNonce).toBe('5');
    });

    it('should update user nonce when changing the address', async () => {
      wrapper.vm.getNextNonce = jest.fn().mockResolvedValueOnce('2');
      wrapper.setData({ address: '0xddddd' });
      await wrapper.vm.$nextTick();

      expect(wrapper.vm.$data.userNonce).toBe('2');

      wrapper.vm.getNextNonce.mockResolvedValueOnce('7');
      wrapper.setData({ address: '0xkdjf' });
      await wrapper.vm.$nextTick();

      expect(wrapper.vm.$data.userNonce).toBe('7');
    });

    it('should call get ENS after each network change', async () => {
      expect.assertions(1);

      wrapper.setData({
        address: 'ether.eth',
      });

      wrapper.vm.getEnsAddress = jest.fn();

      wrapper.setComputed({
        isEnsTransaction: true,
        activeNet: {
          id: 2,
        },
      });
      wrapper.setComputed({
        isEnsTransaction: true,
        activeNet: {
          id: 3,
        },
      });

      expect(wrapper.vm.getEnsAddress).toHaveBeenCalledTimes(2);
    });
  });
});
