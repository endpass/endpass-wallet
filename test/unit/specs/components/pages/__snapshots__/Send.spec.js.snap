// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Send render should render ens error 1`] = `
<div class="app-page send-page">
    <div class="section is-narrow">
        <div class="container is-narrow">
            <div class="card app-card main-app-card">
                <div class="card-header">
                    <h1 class="card-header-title">Send ETH</h1>
                </div>
                <div class="card-content">
                    <v-form id="sendEther" v-on:submit="bound handleTransactionFormSubmit">
                        <div class="field">
                            <label class="label">
                                To
                            </label>
                            <account-chooser creatable="true" width="35" accounts="" placeholder="0x... or ENS" v-model="address"></account-chooser>
                            <!---->
                            <p class="help is-danger">
                                foo
                            </p>
                        </div>
                        <div class="send-amount field is-horizontal">
                            <div class="field-label is-normal">
                                <label for="amount" class="label">
                                    Amount
                                </label>
                            </div>
                            <div class="field-body">
                                <v-input id="value" validator="required|decimal:18|between:0,1" type="number" min="0" step="any" name="value" data-vv-as="amount" aria-describedby="value" placeholder="Amount" required="" v-model="value"><span slot="addon" class="select"><v-select options="[object Object],AAA" name="currencies" v-model="transaction.tokenInfo"></v-select></span>
                                    <a slot="icon" title="Send entire balance" v-on:click="bound setMaxAmount"><span class="icon is-small is-right">{}</span></a>
                                </v-input>
                                <v-input id="price" validator="required|decimal:2|between:0,399.99" type="number" min="0" step="0.01" name="price" aria-describedby="price" placeholder="Price" required="" v-model="price">
                                    <div slot="addon" class="control">
                                        <a class="button is-static">USD</a>
                                    </div>
                                </v-input>
                            </div>
                        </div>
                        <div class="priority-options field is-horizontal">
                            <div class="field-label">
                                <label class="label">Priority</label>
                            </div>
                            <div class="field-body">
                                <v-radio id="priority" options="[object Object],[object Object],[object Object]" name="priority" v-model="transaction.gasPrice"></v-radio>
                            </div>
                        </div>
                        <div class="advanced-options-container">
                            <div class="field advanced-toggle is-horizontal">
                                <div class="field-label"></div>
                                <div class="field-body">
                                    <a v-on:click="bound toggleShowAdvanced" class="has-text-link">
                                        Advanced Options...
                                    </a>
                                </div>
                            </div>
                            <div v-show="showAdvanced" class="advanced-options" style="display: none;">
                                <div class="field is-horizontal">
                                    <div class="field-label">
                                        <label class="label">Gas Price</label>
                                    </div>
                                    <div class="field-body">
                                        <v-input id="gasPrice" name="gasPrice" type="number" min="1" max="100" step="1" validator="required|numeric|integer|between:1,100" aria-describedby="gasPrice" placeholder="Gas price" required="" v-model="transaction.gasPrice">
                                            <div slot="addon" class="control">
                                                <a class="button is-static">Gwei</a>
                                            </div>
                                        </v-input>
                                    </div>
                                </div>
                                <div class="field is-horizontal">
                                    <div class="field-label">
                                        <label class="label">Gas Limit</label>
                                    </div>
                                    <div class="field-body">
                                        <v-input id="gasLimit" name="gasLimit" type="number" min="21000" max="1000000" step="1000" validator="required|numeric|integer|between:21000,4000000" aria-describedby="gasLimit" placeholder="Gas limit" required="" v-model="transaction.gasLimit"></v-input>
                                    </div>
                                </div>
                                <div class="field is-horizontal">
                                    <div class="field-label">
                                        <label class="label">Nonce</label>
                                    </div>
                                    <div class="field-body">
                                        <v-input id="nonce" validator="required|numeric|integer|min_value:0" name="nonce" type="number" step="1" aria-describedby="nonce" placeholder="Nonce" required="" v-on:input="bound setTrxNonce" v-model="userNonce"></v-input>
                                    </div>
                                </div>
                                <div class="field is-horizontal">
                                    <div class="field-label">
                                        <label class="label">Data</label>
                                    </div>
                                    <div class="field-body">
                                        <v-input id="data" name="data" validator="required|hex" aria-describedby="data" placeholder="Data" required="" v-show="!transaction.tokenInfo" v-model="transaction.data"></v-input>
                                    </div>
                                </div>
                            </div>
                        </div>
                        <div class="field is-horizontal">
                            <div class="field-label"></div>
                            <div class="field-body">
                                <v-button disabled="true" class-name="is-success is-medium is-cta" data-test="transaction-send-button">
                                    Send
                                </v-button>
                            </div>
                        </div>
                    </v-form>
                    <!---->
                </div>
            </div>
        </div>
    </div>
    <!---->
    <!---->
</div>
`;

exports[`Send render should render resolved ens address 1`] = `
<div class="app-page send-page">
    <div class="section is-narrow">
        <div class="container is-narrow">
            <div class="card app-card main-app-card">
                <div class="card-header">
                    <h1 class="card-header-title">Send ETH</h1>
                </div>
                <div class="card-content">
                    <v-form id="sendEther" v-on:submit="bound handleTransactionFormSubmit">
                        <div class="field">
                            <label class="label">
                                To
                            </label>
                            <account-chooser creatable="true" width="35" accounts="" placeholder="0x... or ENS" v-model="address"></account-chooser>
                            <p class="help ellipsis">
                                Resolved ENS address: 0x0
                            </p>
                            <!---->
                        </div>
                        <div class="send-amount field is-horizontal">
                            <div class="field-label is-normal">
                                <label for="amount" class="label">
                                    Amount
                                </label>
                            </div>
                            <div class="field-body">
                                <v-input id="value" validator="required|decimal:18|between:0,1" type="number" min="0" step="any" name="value" data-vv-as="amount" aria-describedby="value" placeholder="Amount" required="" v-model="value"><span slot="addon" class="select"><v-select options="[object Object],AAA" name="currencies" v-model="transaction.tokenInfo"></v-select></span>
                                    <a slot="icon" title="Send entire balance" v-on:click="bound setMaxAmount"><span class="icon is-small is-right">{}</span></a>
                                </v-input>
                                <v-input id="price" validator="required|decimal:2|between:0,399.99" type="number" min="0" step="0.01" name="price" aria-describedby="price" placeholder="Price" required="" v-model="price">
                                    <div slot="addon" class="control">
                                        <a class="button is-static">USD</a>
                                    </div>
                                </v-input>
                            </div>
                        </div>
                        <div class="priority-options field is-horizontal">
                            <div class="field-label">
                                <label class="label">Priority</label>
                            </div>
                            <div class="field-body">
                                <v-radio id="priority" options="[object Object],[object Object],[object Object]" name="priority" v-model="transaction.gasPrice"></v-radio>
                            </div>
                        </div>
                        <div class="advanced-options-container">
                            <div class="field advanced-toggle is-horizontal">
                                <div class="field-label"></div>
                                <div class="field-body">
                                    <a v-on:click="bound toggleShowAdvanced" class="has-text-link">
                                        Advanced Options...
                                    </a>
                                </div>
                            </div>
                            <div v-show="showAdvanced" class="advanced-options" style="display: none;">
                                <div class="field is-horizontal">
                                    <div class="field-label">
                                        <label class="label">Gas Price</label>
                                    </div>
                                    <div class="field-body">
                                        <v-input id="gasPrice" name="gasPrice" type="number" min="1" max="100" step="1" validator="required|numeric|integer|between:1,100" aria-describedby="gasPrice" placeholder="Gas price" required="" v-model="transaction.gasPrice">
                                            <div slot="addon" class="control">
                                                <a class="button is-static">Gwei</a>
                                            </div>
                                        </v-input>
                                    </div>
                                </div>
                                <div class="field is-horizontal">
                                    <div class="field-label">
                                        <label class="label">Gas Limit</label>
                                    </div>
                                    <div class="field-body">
                                        <v-input id="gasLimit" name="gasLimit" type="number" min="21000" max="1000000" step="1000" validator="required|numeric|integer|between:21000,4000000" aria-describedby="gasLimit" placeholder="Gas limit" required="" v-model="transaction.gasLimit"></v-input>
                                    </div>
                                </div>
                                <div class="field is-horizontal">
                                    <div class="field-label">
                                        <label class="label">Nonce</label>
                                    </div>
                                    <div class="field-body">
                                        <v-input id="nonce" validator="required|numeric|integer|min_value:0" name="nonce" type="number" step="1" aria-describedby="nonce" placeholder="Nonce" required="" v-on:input="bound setTrxNonce" v-model="userNonce"></v-input>
                                    </div>
                                </div>
                                <div class="field is-horizontal">
                                    <div class="field-label">
                                        <label class="label">Data</label>
                                    </div>
                                    <div class="field-body">
                                        <v-input id="data" name="data" validator="required|hex" aria-describedby="data" placeholder="Data" required="" v-show="!transaction.tokenInfo" v-model="transaction.data"></v-input>
                                    </div>
                                </div>
                            </div>
                        </div>
                        <div class="field is-horizontal">
                            <div class="field-label"></div>
                            <div class="field-body">
                                <v-button class-name="is-success is-medium is-cta" data-test="transaction-send-button">
                                    Send
                                </v-button>
                            </div>
                        </div>
                    </v-form>
                    <!---->
                </div>
            </div>
        </div>
    </div>
    <!---->
    <!---->
</div>
`;
